src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseHeaderMap.cs

```diff
@@ -116,91 +112,90 @@ private void ValidateIsReadOnly()
-
-                if (label.LabelAsString == null)
-                {
-                    // all known headers are integers.
-                    ValidateKnownHeaderValue(label.LabelAsInt32, reader.PeekState(), reader);
-                }
-                else
-                {
-                    reader.SkipValue();
-                }
+                CborReaderState initialState = reader.PeekState();
+                int intLabel = label.LabelAsInt32;
 
-                if (reader.BytesRemaining != 0)
-                {
-                    throw new InvalidOperationException(SR.Format(SR.CoseHeaderMapCborEncodedValueNotValid, label));
-                }
-            }
-
-            static void ValidateKnownHeaderValue(int label, CborReaderState? initialState, CborReader? reader)
-            {
-                switch (label)
+                switch (intLabel)
                 {
                     case KnownHeaders.Alg:
                         if (initialState != CborReaderState.NegativeInteger &&
                             initialState != CborReaderState.UnsignedInteger &&
                             initialState != CborReaderState.TextString)
                         {
-                            throw new InvalidOperationException(SR.Format(SR.CoseHeaderMapHeaderDoesNotAcceptSpecifiedValue, label));
+                            throw new InvalidOperationException(SR.Format(SR.CoseHeaderMapHeaderDoesNotAcceptSpecifiedValue, intLabel));
                         }
-                        reader?.SkipValue();
+                        reader.SkipValue();
                         break;
                     case KnownHeaders.Crit:
-                        reader?.SkipValue(); // TODO
-                        break;
-                    case KnownHeaders.CounterSignature:
-                        reader?.SkipValue(); // TODO
+                        int length = reader.ReadStartArray().GetValueOrDefault();
+                        if (length < 1)
+                        {
+                            throw new InvalidOperationException(SR.CriticalHeadersMustBeArrayOfAtLeastOne);
+                        }
+
+                        for (int i = 0; i < length; i++)
+                        {
+                            CborReaderState state = reader.PeekState();
+                            if (state == CborReaderState.UnsignedInteger || state == CborReaderState.NegativeInteger)
+                            {
+                                reader.ReadInt32();
+                            }
+                            else if (state == CborReaderState.TextString)
+                            {
+                                reader.ReadTextString();
+                            }
+                            else
+                            {
+                                throw new InvalidOperationException(SR.Format(SR.CoseHeaderMapHeaderDoesNotAcceptSpecifiedValue, intLabel));
+                            }
```

Comment:
More concern around the exception model of this method, as a mix of CBOR and InvalidOperation exceptions when it's all "your argument is bad".